Harbor replication issues since upgrade to OSS version:v2.1.3

What can we help you?

We're running Harbor on Kubernetes v1.18.15. We recently upgraded Harbor from Helm chart v1.4.6 (Harbor OSS version:v2.0.6) to Helm chart v1.5.3 (Harbor OSS version:v2.1.3)

Since the upgrade, we've noticed a number of errors re replication:

- One of the replication rules can't be disabled, we get an `Internal Server Error` in the UI, which on the `jobservice` logs translates to:
```
[ERROR] [/jobservice/api/handler.go:288]: Serve http request 'POST /api/v1/jobs/d353946f090203dbc8ca5035' error: 404 {"code":10010,"message":"object is not found","details":"d353946f090203dbc8ca5035"}
```

- Another replication rule, which is currently trying to replicate artifacts to a destination Harbor instance, and uses a robot account for authentication, errors out complaining about said robot account not existing:

```
failed to do the prepare work for pushing/uploading resources: http error: code 404, message {"errors": [{"code":"NOT_FOUND","message":"user robot$XXXXX not found"
```

... but the robot account does exist, and this was working fine before.

In terms of configuration, we're currently using external components for Harbor database (RDS Postgres), chart/image persistence (S3 bucket) and caching layer (Redis Elasticache).
I've tried stuff such as Harbor restarts, Redis flushing, GC runs... to no avail.

@cparadal 

When meeting the error you mentioned, is it possible to check the reids data if the job stats is still there?

>Since the upgrade, we've noticed a number of errors re replication:
One of the replication rules can't be disabled, we get an Internal Server Error in the UI, which on the jobservice logs translates to:

`[ERROR] [/jobservice/api/handler.go:288]: Serve http request 'POST /api/v1/jobs/d353946f090203dbc8ca5035' error: 404 {"code":10010,"message":"object is not found","details":"d353946f090203dbc8ca5035"}`


The key of the  job stats is {js-namespace}:job_stats:<job_id>.

The job stats might be gone as there is a default expiration time for reducing the data kept in the Redis DB.

We're also looking at the code at core side to see if it handles the 404 case correctly.

--- a/src/core/api/admin_job.go
+++ b/src/core/api/admin_job.go
@@ -65,13 +65,11 @@ func (aj *AJAPI) updateSchedule(ajr models.AdminJobReq) {
 
 	// stop the scheduled job and remove it.
 	if err = utils_core.GetJobServiceClient().PostAction(jobs[0].UUID, common_job.JobActionStop); err != nil {
-		_, ok := err.(*common_job.StatusBehindError)
-		if !ok {
-			if e, ok := err.(*common_http.Error); !ok || e.Code != http.StatusNotFound {
-				aj.SendInternalServerError(err)
-				return
-			}
+		if err != common_job.ErrJobNotFound {
+			aj.SendInternalServerError(err)
+			return
 		}
+		log.Warningf("the job %s not found on jobservice, skip directly", jobs[0].UUID)
 	}
 
 	if err = dao.DeleteAdminJob(jobs[0].ID); err != nil {

--- a/src/replication/policy/scheduler/scheduler.go
+++ b/src/replication/policy/scheduler/scheduler.go
@@ -16,11 +16,8 @@ package scheduler
 
 import (
 	"fmt"
-	"net/http"
-	"strings"
 	"time"
 
-	commonHttp "github.com/goharbor/harbor/src/common/http"
 	"github.com/goharbor/harbor/src/common/job"
 	jobModels "github.com/goharbor/harbor/src/common/job/models"
 	jsJob "github.com/goharbor/harbor/src/jobservice/job"
@@ -103,13 +100,12 @@ func (s *scheduler) Unschedule(policyID int64) error {
 	}
 	for _, sj := range sjs {
 		if err = s.jobservice.PostAction(sj.JobID, job.JobActionStop); err != nil {
-			// if the job specified by jobID is not found in jobservice, just delete
-			// the record from database
-			if e, ok := err.(*commonHttp.Error); !ok || (e.Code != http.StatusNotFound &&
-				!strings.Contains(e.Message, "no valid periodic job policy found")) {
+			if err != job.ErrJobNotFound {
 				return err
 			}
-			log.Debugf("the stop action for schedule job %s submitted to the jobservice", sj.JobID)
+			// if the job specified by jobID is not found in jobservice, just delete
+			// the record from database
+			log.Warningf("the job %s not found on jobservice, skip directly", sj.JobID)
 		}
 		if err = dao.ScheduleJob.Delete(sj.ID); err != nil {
 			return err

